[GStreamer] Clang warns  about returning stack value in GStreamerPeerConnectionBackend.cpp

aperezdc opened the pull request:
#### bf68b55937f3bf63510b133f01ea3f1ab1b467d6
<pre>
[GStreamer] Clang warns  about returning stack value in GStreamerPeerConnectionBackend.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=300616">https://bugs.webkit.org/show_bug.cgi?id=300616</a>

Reviewed by Philippe Normand.

The newly created transceiver is kept alive by the RTCPeerConnection
after the call to RTCPeerConnection::addInternalTransceiver(), and
therefore it is valid to return its address. The added annotations
to the Ref template in 301385@main make Clang notice that, indeed a
reference to a stack-local variable is returned because Ref::get()
does exactly that. Returning the pointer directly using Ref::ptr()
sidesteps the issue, and avoid needing to apply the &amp; operator later
on. While at at it, move the Ref into the addInternalTransceiver()
call to avoid reference count churn.

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::connectIncomingTrack):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::newRemoteTransceiver):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:

Canonical link: <a href="https://commits.webkit.org/301402@main">https://commits.webkit.org/301402@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/551e1363ee5208bf285c57a69ede84863d0ec5c7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125897 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45559 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36314 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132767 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77761 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127768 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46242 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54117 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95915 "Found 3 new test failures: imported/w3c/web-platform-tests/event-timing/crossiframe.html imported/w3c/web-platform-tests/html/browsers/the-window-object/window-open-windowfeatures-values.html imported/w3c/web-platform-tests/webrtc-extensions/RTCRtpTransceiver-headerExtensionControl.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64015 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128845 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36980 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112578 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76406 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35886 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30762 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76239 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106762 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30975 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135450 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52682 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40416 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104384 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53131 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108788 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104111 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26506 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49482 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27796 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50053 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52576 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58388 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51919 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55270 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53618 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
```diff
@@ -426,8 +426,9 @@ RTCRtpTransceiver& GStreamerPeerConnectionBackend::newRemoteTransceiver(std::uni
     auto transceiver = RTCRtpTransceiver::create(WTFMove(sender), WTFMove(receiver), WTFMove(transceiverBackend));
-    peerConnection->addInternalTransceiver(transceiver.copyRef());
-    return transceiver.get();
+    auto* result = transceiver.ptr();
+    peerConnection->addInternalTransceiver(WTFMove(transceiver));
+    return result;
```
Even if this is fine now, I am worried this might lead to bugs down the line.

The logic here relies on `addInternalTransceiver` keeping the `transceiver` alive in a specific way that does not invalidate the pointer stored in result. But that's an implementation detail that is true today -- in the future, it might not.

If clients using `newRemoteTransceiver` should participate in the transceiver lifecycle, the might need to get a RefPtr themselves. If they don't participate, they could use a weak pointer. But a raw pointer of a moved-from object feels tricky...

aperezdc replied:
I understand the concern, but the reliance on `RTCPeerConnection` keeping the transceiver alive was already the situation before.

If we want to change `::newRemoteTransceiver()` to return a `RefPtr`, then we also need to change `::existingTransceiver()` and the code that calls them. I am not against doing that, but IMO it would better be done in a separate PR&mdash;the intention with this one is to only get the build going.

@philn: What do you think? Is it okay to have a bit of extra reference count churn in the code paths that use `::existingTransceiver()` and `::newRemoteTransceiver()` if we change them in a followup PR to return `RefPtr`? My understanding is that this is not hot code, but I would prefer somebody with more multimedia chops to comment on that.

philn replied:
A follow-up sounds good to me!

aperezdc replied:
Great, landing this, and filed [issue #300625](https://bugs.webkit.org/show_bug.cgi?id=300625) for the follow-up.

aperezdc committed:
[GStreamer] Clang warns  about returning stack value in GStreamerPeerConnectionBackend.cpp
https://bugs.webkit.org/show_bug.cgi?id=300616

Reviewed by Philippe Normand.

The newly created transceiver is kept alive by the RTCPeerConnection
after the call to RTCPeerConnection::addInternalTransceiver(), and
therefore it is valid to return its address. The added annotations
to the Ref template in 301385@main make Clang notice that, indeed a
reference to a stack-local variable is returned because Ref::get()
does exactly that. Returning the pointer directly using Ref::ptr()
sidesteps the issue, and avoid needing to apply the & operator later
on. While at at it, move the Ref into the addInternalTransceiver()
call to avoid reference count churn.

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::connectIncomingTrack):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::newRemoteTransceiver):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:

Canonical link: https://commits.webkit.org/301402@main
